### PR TITLE
[Sol->Yul] Implementing type name.

### PR DESCRIPTION
**Pull request by mijovic**
Part of #10390

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Don't we have some kind of helper for storing literal data? This case it should just go into a `data` section in the object.

The helper should just decide based on length though whether to keep it in code or as a data section (and for contract names, most likely in-code is the better option).

Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I didn't know it exists, but let me try searching for it. I checked in `revertStringifDebug` it was done same way.

**Comment by mijovic** [comment]
`yul_unimplemented` fails now. It is going to be a bit harder finding something unimplemented this time :-)

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Regardless of how we store the string, we should have a yul util function to call with a literal string that returns a newly allocated memory array containing it.

File: test/cmdlineTests/yul_unimplemented/err
Comment (axic) [inline test/cmdlineTests/yul_unimplemented/err:1]: This test is now useless (it is called `yul_unimplemented`). Can you come up with something new? Or if not, then just delete it.

Comment (mijovic) [reply test/cmdlineTests/yul_unimplemented/err:1]: I wanted to wait for the rest of my Sol->Yul PRs to be merged to come up with something for this test... Will try maybe before merging them if I fond something.

**Comment by chriseth** [comment]
Maybe it's not really worth having that test anymore...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I moved storing string to util function.
Couldn't find helper that stores it to data section. Probably we can do that as part of other PR if needed.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Please use some kind of `define` helper function.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Can you re-use this in other places?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: `_literal` has to be escaped somehow - or use the hash.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I will use hash, fixing now

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Fixing

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I think revert reason if debug has the same code like this helper. Will try using it over there.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Hm, it can be reused in there, but I think that piece of code in revert reason if debug can't camm other functions (as it is called from inline assembly at some places). It was like that before, I'll need to re-check that.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: There's also: abiEncodingFunctionStringLiteral YulUtilFunctions::conversionFunctionSpecial and YulUtilFunctions::revertReasonIfDebug

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Yes, revertReasonIfDebug is inlined in certain places, so maybe let's keep that as it is.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: On the other hand, changing that to a function might also make sense at some point. It cannot be changed to a function everywhere, though.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: For `revertReasonIfDebug` we have #8456 and once it is done, we should use this helper. Also, this would leave link there as a reminder 
I will update this PR with usage in other functions now.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Needed to refactor a bit, and called this function from `abiEncodingFunctionStringLiteral` and `conversionFunctionSpecial`

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Can use `arrayAllocationFunction`, which also stores the size.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: This does not compute a hash and will lead to a collision for strings that are the same in the first 256 bits.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Same here.

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Can you check with the name simplifier that the last part is properly cut off here if it is unambiguous?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Do we have some helper function to calculate hash, or can I use `std::hash`?

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I used same method for calculating `richIdentifier` in `StringLiteralType`. Should be fine now I guess

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Why not name this `copyLiteralToMemory`?

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Fixed now

**Comment by mijovic** [comment]
Updated

File: libsolidity/codegen/YulUtilFunctions.h
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	/// @returns the name of a function that copies a string literal to memory
```
I think it's clearer that way, like the other docs

File: libsolidity/codegen/YulUtilFunctions.h
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	/// and returns a pointer to the memory area containing the string literal.
```
is my suggestion correct?

File: libsolidity/codegen/YulUtilFunctions.h
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	/// @returns the name of a function that stores a string literal at a specific location in memory
```

File: libsolidity/codegen/YulUtilFunctions.h
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]: A bit pedantic, but `store ... to` sounds a bit weird to me, I'd replace it to `storeLiteralInMemoryFunction`

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: It is correct, will update

File: libsolidity/codegen/YulUtilFunctions.h
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: Changed

**Comment by mijovic** [comment]
@leonardoalt Should be updated with all suggestions now

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (axic) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
	string functionName = "store_literal_in_memory_" + util::toHex(util::keccak256(_literal).asBytes());
```

I'm not sure if we ensure that the yul helper function name matches the c++ one, but maybe it is a good idea to do so.

**Review by leonardoalt — DISMISSED** [review]

**Comment by mijovic** [comment]
@leonardoalt Can you re-approve, I fixed cmdLine test that was failing due to renaming yul util function

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**
